Resolve most C++11 "TODO" comments

peter-b opened the pull request:
Now that we're building with MSVC 2015, nearly all the `TODO` comments relating to C++11 support or lack thereof can be fixed.

Mostly this boils down to:

- Adding `constexpr` constructors
- Converting some `inline` functions to `constexpr`
- Adding static initialisation to member variables

Small annoyance: C-style bitfields (used widely in the engine for `bool` member variables) are incompatible with static initialisation, so any `struct` or `class` that has bitfield member variables have to have a `constexpr` default constructor. Sad!

runrevmark commented:
@peter-b: Is there another approach to the bitfield type thing which would solve that problem? Could the bitfields be expressed as a distinct type for example?

peter-b commented:
@runrevmark I looked into it for a while and it appears that it's just not possible.

peter-b committed:
libfoundation: Make MCDataGetSpan<>() default to byte_t

`MCDataGetSpan<>()` can now be used without a type parameter, in which
case it will return an `MCSpan<byte_t>`.

peter-b committed:
libgraphics: Add static initialisers to many structures

peter-b committed:
Fix missing inclusion prevent correct destructor use

peter-b committed:
libfoundation: Use C++11 `default` constructors & static initializers

peter-b committed:
libfoundation: Give all managed pointer types constexpr constructors

Allow aggregate types that include libfoundation managed pointers to
have `constexpr` constructors.

peter-b committed:
libfoundation: Require compilers to support C++11 constexpr

Remove workaround code for compiling LiveCode on C++ compilers with no
C++11 `constexpr` support.  N.b. we don't require C++14 generalised
`constexpr` yet, so `constexpr` functions must contain exactly one
return statement.

peter-b committed:
Use `std::swap` to move construct/assign object handles

peter-b committed:
MCPendingMessage: Use trivial constructors

The default constructor and copy constructor for `MCPendingMessage`
are trivial, and the default constructor can be `constexpr`.

peter-b committed:
Allow object handles to be `constexpr` default constructed

This allows any `class` or `struct` that encapsulates an object handle
to have a `constexpr` default constructor.

peter-b committed:
libfoundation: Make byte-swapping functions constexpr

peter-b committed:
libgraphics: Remove bitfields from structures

In all of the structures described in the `graphics.h` header, using
bitfields had no effect on structure length or layout, while
preventing the use of static initialisation.  @runrevmark recommended
to remove the bitfields entirely.

runrevmark commented:
@livecode-vulcan review ok 952685a

livecode-vulcan commented:
💙 review by @runrevmark ok 952685a0fddd90b6b160758d6f1c8d57329a61b2

livecode-vulcan commented:
😞 test failure 952685a0fddd90b6b160758d6f1c8d57329a61b2

* try-community-linux-x86: [failure](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1063)
* try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/1066)
* try-community-linux-x86_64: [failure](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1058)
* try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1064)
* try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/1068)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1080)
* try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1058)

peter-b committed:
libfoundation: Declare MCSpan copy constructor correctly

Copy constructors need to take a `const` reference as their only
argument.

runrevmark commented:
@livecode-vulcan review ok 5028612

livecode-vulcan commented:
💙 review by @runrevmark ok 502861296af35c75fbaa41d507cc58bea7003406

livecode-vulcan commented:
😞 test failure 502861296af35c75fbaa41d507cc58bea7003406

* try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1064)
* try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/1067)
* try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1059)
* try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1065)
* try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/1069)
* try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1081)
* try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1059)

peter-b commented:
@livecode-vulcan retry 5028612

livecode-vulcan commented:
😞 test failure 502861296af35c75fbaa41d507cc58bea7003406

* try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1065)
* try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/1068)
* try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1060)
* try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1066)
* try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/1070)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1082)
* try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1060)

peter-b committed:
Make narrowing conversions explicit in MCGSize list initialisation

montegoulding commented:
So should we discourage the use of bitfields in new code? I'm asking because I have a heap in the platform api.

peter-b commented:
@montegoulding I would normally recommend not to use bitfields unless there is a specific memory layout problem that you are trying to solve.

montegoulding commented:
I think the only memory layout problems I'm working on are in my brain @peter-b ;-)

peter-b commented:
@montegoulding Tell me about it 😢

peter-b commented:
In the last version of this PR, it turned out that including `<algorithm>` (for `std::swap`) was interfering the `<math.h>` somehow. ⁉️  I've changed the patch to use `<utility>` instead.

livecodeali commented:
@livecode-vulcan review ok 1ea65ed

livecode-vulcan commented:
💙 review by @livecodeali ok 1ea65ed77d636d4934b172ef81bbbf8c334967d6

livecode-vulcan commented:
😎 test success 1ea65ed77d636d4934b172ef81bbbf8c334967d6

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1076)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1079)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1071)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1077)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1081)
* try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1093)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1072)
